### PR TITLE
[Autoscaling] Add Elasticsearch autoscaling controller

### PR DESCRIPTION
**Pull request by barkbay**
This PR introduces the Elasticsearch autoscaling controller.

This new controller manages the autoscaling policies, periodically polls the `_autoscaling/capacity` API and adjusts automatically the resources in the node sets.

Some parts of the code are not optimized, this PR is large, I tried to keep something which is easy to understand.
I also tried to add comments where needed to help the reader/reviewer.

Starting by the `reconcileInternal` function in `pkg/controller/autoscaling/elasticsearch/driver.go` might be a good starting point for a review.

The following improvements will be added in dedicated PRs:

- [ ] Add a warning in case of overlapping roles.
- [ ] Manage limits using ratios, only resource requests are supported for now.
- [ ] Add a stabilization period
- [ ] Make the poll period a setting, current default is 60s

## Testing

This PR relies on the [ES autoscaling client](https://github.com/elastic/cloud-on-k8s/pull/4158) and requires to activate a trial (or a valid license)

Deciders are documented [here](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/autoscaling-deciders.html).

### Data tier

Unfortunately I don't have strong advices for testing storage deciders.
You can try to use Rally, but note that it will only exercise the reactive decider:

```
esrally race --pipeline=benchmark-only --track=eventdata --track-repository=eventdata --challenge=index-logs-fixed-daily-volume \
             --track-params="bulk_size:100,number_of_replicas:1,bulk_indexing_clients:1,daily_logging_volume:1GB" \
             --target-hosts=https://${ES}:9200 \
             --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'elastic',basic_auth_password:$PASSWORD"
```

You can also deploy some agents to create datastreams.

### ML

You can use [the following API call](https://gist.githubusercontent.com/barkbay/d956d408366da14e74108b8a22205091/raw/9ca9a95d6d003a0d4ceb9b25e5e5bc0f14d16cfc/Create%2520ML%2520Job) to create a ML job _(it requires the sample data flights)_. Then do a `POST _ml/data_frame/analytics/mljob/_start` to start the job.

## Logs and status

The expected resources for the node sets and some useful messages are printed in the logs but also stored in the `elasticsearch.alpha.elastic.co/autoscaling-status` annotation. You can use the following command to retrieve it:`kubectl get -n elasticsearch.elasticsearch.k8s.elastic.co/mycluster -o jsonpath='{.metadata.annotations.elasticsearch\.alpha\.elastic\.co/autoscaling-status}' | jq`

### CODE REVIEW

**Review by pebrc — COMMENTED** [review]
I am still working through this but I thought I share a first set of comments. All minor.

File: pkg/controller/autoscaling/elasticsearch/status/status.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/status/status.go:28]: I'll fix the comment. I'm reusing `policies` here to match the root `policies` [element in the spec](https://github.com/elastic/cloud-on-k8s/blob/fa7f098583f84888044813d257d5b0ff4500271e/config/recipes/autoscaling/elasticsearch.yaml#L9).

For example, the following spec:

```json
{
	"policies": [{
		"name": "di",
		"roles": ["data", "ingest", "transform"],
		"resources": {
			"nodeCount": { "min": 3, "max": 8},
			"cpu": {"min": 2,"max": 8},
			"memory": {"min": "2Gi","max": "16Gi"},
			"storage": {"min": "64Gi","max": "512Gi"}
		}
	}]
}
```

Might lead to the following status:

```json
{
	"policies": [{
		"name": "di",
		"nodeSets": [{ "name": "nodeset-di-a", "nodeCount": 4}, {"name": "nodeset-di-b","nodeCount": 3}],
		"resources": {
			"requests": {"cpu": "6","memory": "10Gi","storage": "4Gi"}
		},
		"state": [{
			"type": "VerticalScalingLimitReached",
			"messages": [
				"Node required storage 283970435 is greater than max allowed: 209715200"
			]
		}],
		"lastModificationTime": "2021-01-06T09:55:57Z"
	}]
}
```

Reusing `policies` helps to have a consistency between the top-level structure of the 2 documents:

```
{
	"policies": [{
		"name": "di", ...
}
```

File: pkg/controller/autoscaling/elasticsearch/status/status.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/status/status.go:45]: I shouldn't have used the term "named tier", I agree it is confusing. I'll fix that.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go
Comment (charith-elastic) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go:44]: Could we sort the list once and then treat it like a circular queue? I.e. maintain the current "head" of the array and move it to the next element (wrapping around if necessary) with each call to `AddNode`? 

We could also consider using https://golang.org/pkg/container/heap/ but that's probably overkill for this use case.

File: pkg/controller/elasticsearch/driver/autoscaling.go
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/autoscaling.go:40]: `s` is confusing me. Should it be `r` for resources? Or even `expectedResources`

File: pkg/controller/elasticsearch/driver/autoscaling.go
Comment (pebrc) [inline pkg/controller/elasticsearch/driver/autoscaling.go:47]:
```suggestion
		inSync, err := expectedResources.Match(esv1.ElasticsearchContainerName, nodeSet)
```
?

File: pkg/controller/autoscaling/elasticsearch/driver.go
Comment (charith-elastic) [inline pkg/controller/autoscaling/elasticsearch/driver.go:167]: Nit: Could this be an `if/else`?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go:45]:
```suggestion
	// Peak the first element, this is the one with fewer nodes
```

**Review by charith-elastic — COMMENTED** [review]
It's a lot to digest and I won't claim to have understood everything. But, overall LGTM 👍🏼

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go:38]: Should we allow CPU capacity in smaller increments than one core?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go:56]: max value?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go:64]: max value?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler_test.go:72]: max value?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go:55]: This abstraction strikes me as a bit odd. Could this not simply be a function?
```golang
distributeFairly(nodeSetsResources, expectedNodeCount)
```

Also: the logger seems unused.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go:69]:
```suggestion
// If user removed the limits while offline we are assuming that they want to take back control of the resources.
```

File: pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go:71]: Why are we passing the status here once more we already have extracted the `currentNodeSetsResouces` from the status which should contain everything we need?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go:121]:
```suggestion
// adjustQuantity ensures that the Quantity in value is between min and max.
```

**Review by pebrc — COMMENTED** [review]
I think I am 2/3 through ...

File: pkg/controller/autoscaling/elasticsearch/controller.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/controller.go:113]: Good point, since I think we still want to check the compatibility I introduced a dedicated function in https://github.com/elastic/cloud-on-k8s/pull/4197

File: pkg/controller/autoscaling/elasticsearch/resources/resources_test.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/resources/resources_test.go:116]: Test name out of sync

File: pkg/controller/autoscaling/elasticsearch/resources/resources.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/resources/resources.go:118]: Would it makes sense to have this as a map[string]int32 to begin with instead of having a custom struct and converting to maps on the fly?

File: pkg/controller/autoscaling/elasticsearch/status/events.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/status/events.go:25]: Just curious why are we dropped certain event types here?

**Review by pebrc — COMMENTED** [review]
Impressive work! I think I went through everything once. But I haven't tested as this is depending on the API PR IIUC

File: pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go]: You're right, done in 3f462982f

File: pkg/controller/autoscaling/elasticsearch/controller.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/controller.go:46]: 👍  for `defaultRequeue`, this is also what we use in various [other](https://github.com/elastic/cloud-on-k8s/blob/fa7f098583f84888044813d257d5b0ff4500271e/pkg/controller/association/reconciler.go#L40) [places](https://github.com/elastic/cloud-on-k8s/blob/fa7f098583f84888044813d257d5b0ff4500271e/pkg/controller/elasticsearch/driver/driver.go#L46).

File: pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/nodesets.go]: I think this is included in "Some parts of the code are not optimized" 😄 
I needed to have something that "just" do the work to make some progress on others, more important, parts. As long as it works as expected I think I'm happy to let someone else optimize it in an other PR.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go:38]: I don't have a strong opinion. I always have preferred rounded values because it makes it easier to resonate about the performance model (e.g. how to size thread pools in the JVM with fractional number ?) But I agree that the user might want to have more accurate resource requests.
@elastic/cloud-k8s I'm curious if there are other opinions.

File: pkg/controller/autoscaling/elasticsearch/reconcile.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/reconcile.go:104]: Good catch ! 🤦

File: pkg/controller/autoscaling/elasticsearch/resources/resources.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/resources/resources.go:118]: The aim of the `NodeSetNodeCount` struct is to make it easier to serialize the status with the pattern `{ name: "...", .... }`, for example:

```
	"nodeSets": [
        {
		"name": "nodeset-di-a",
		"nodeCount": 4
	},
        {
		"name": "nodeset-di-b",
		"nodeCount": 3
	}]
```

I wanted to be consistent with other items in the specs and explicitly have a `name` field in the status.

File: pkg/controller/autoscaling/elasticsearch/status/status.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/status/status.go:125]: This is something I didn't try to optimize yet. The API has been designed to be used this way, but I agree we should try to be clever here. Since this is an optimization I would track it with a dedicated issue and a separate PR ?

File: pkg/controller/autoscaling/elasticsearch/policy.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/policy.go:27]: This is something I didn't try to optimize yet. The API has been designed to be used this way, but I agree we should try to be clever here. Since this is an optimization I would track it with a dedicated issue and a separate PR ?

File: pkg/controller/autoscaling/elasticsearch/autoscaler/horizontal.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/horizontal.go:64]: This is actually the number of nodes that must be added to fulfill the requirements, I tried to improve the comment [here](https://github.com/elastic/cloud-on-k8s/pull/4173/files#diff-e0e0ede710eb10e3e4b3e10b8aa6f25802dc7a18a4784fa66b16410613de6b4dR21).

For example, if there is a minimum number of 2 (because this is what the user specified in the spec) but we need a total of 5 nodes in the cluster then `nodeToAdd` is 3

File: pkg/controller/autoscaling/elasticsearch/status/events.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/status/events.go:25]: I believe my initial thinking was that there may not be a 1-1 relationship between the events we want in the resource status _(primarily targeting the user)_ and the events we want to raise at the cluster level _(targeting the user and the K8S cluster admins)_.
For example I'm not sure that an empty response from Elasticsearch (`EmptyResponse`  event), which is likely to be a temporary situation, really deserves a K8S cluster event. While reaching a limit, on the contrary, is a critical situation which might require to be surfaced more broadly.
I agree it is not a strong argument, just a feeling, so I think I'm fine removing this filter for the moment.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/autoscaler_test.go
Comment (barkbay) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/autoscaler_test.go:124]: Done in https://github.com/elastic/cloud-on-k8s/commit/e96a25e3eb591f5e4deee999c3f6b2dac89c70e3

File: pkg/controller/autoscaling/elasticsearch/status/events.go
Comment (pebrc) [reply pkg/controller/autoscaling/elasticsearch/status/events.go:25]: Makes sense, we can adjust this later as well.

File: pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go
Comment (mapdegree) [reply pkg/controller/autoscaling/elasticsearch/autoscaler/linear_scaler.go:38]: For Elasticsearch or Kibana, we would round up CPU capacity to an integer, but might be too much for lighter components, for example beats. 
Also, if you just want to run the whole ECK stack on a laptop, users would prefer to have less granual option to define resource limits, just to enable them to test locally.

**Comment by barkbay** [comment]
Thanks all for the time you spent on this PR, I know it is not an easy one 🙇 

Quick update:
* I have been working on a [PR to support request/limit ratio](https://github.com/barkbay/cloud-on-k8s/compare/autoscaling-pr/controller...barkbay:tmp/autoscaling-pr/limits) but I think it requires this one to be merged before I raise it.
* Also this PR is blocked by this [one](https://github.com/elastic/cloud-on-k8s/pull/4197).

File: pkg/controller/autoscaling/elasticsearch/resources/resources.go
Comment (pebrc) [inline pkg/controller/autoscaling/elasticsearch/resources/resources.go:40]:
```suggestion
// It also returns false if the container is not found in the NodeSet.
```
?

**Review by pebrc — APPROVED** [review]
LGTM! I did only a very few tests over night and storage autoscaling worked as expected. But I hoped to be able to do some more tests. But I believe we can merge this as is and iterate if necessary.

**Pull request merged**
